numpy/core/src/multiarray/getset.c

```diff
@@ -67,28 +67,32 @@ array_shape_set(PyArrayObject *self, PyObject *val)
         return -1;
     }
 
-    /* Free old dimensions and strides */
-    npy_free_cache_dim_array(self);
     nd = PyArray_NDIM(ret);
-    ((PyArrayObject_fields *)self)->nd = nd;
     if (nd > 0) {
         /* create new dimensions and strides */
-        ((PyArrayObject_fields *)self)->dimensions = npy_alloc_cache_dim(2 * nd);
-        if (PyArray_DIMS(self) == NULL) {
+        npy_intp *_dimensions = npy_alloc_cache_dim(2 * nd);
+        if (_dimensions == NULL) {
             Py_DECREF(ret);
             PyErr_SetString(PyExc_MemoryError,"");
```

Comment:
While you're here, may as well change to
```suggestion
            PyErr_NoMemory();
```

Reply:
Maybe worth a quick sweep of all our code for this simplification. In a separate PR.